Wrap all the various way of doing checkouts in GTCheckoutOptions

tiennou opened the pull request:
Fixes #457

tiennou commented:
I hit some "disappointment" while transitioning the clone API. I'm not sure what to do, so I wrote a wrapper for the current BOOL option, but I'd like some feedback on that.

Tests are passing here, and right now it handles the same thing as before, but now it can be more easily extended for the other use cases. I *might* take a look this week-end at adding path support for the checkout case (since I think clone is pretty complete).

@jspahrsummers Feel free to hint me at some use cases you need to switch to Objectified glory ;-).

joshaber commented:
I don't feel strongly about this one way or the other. Anyone else?

mdiep commented:
I'd rather not wrap these in a class, personally.

jspahrsummers commented:
I think this approach makes sense if there are many different operations that may need “checkout options.” I don't know if that's the case or not.

tiennou commented:
Here's a list : [cherrypick](https://github.com/libgit2/libgit2/blob/c03e8c224cf18104bc97a68aed85317820d062ae/include/git2/cherrypick.h#L33), clone (we're handling that already), [merge](https://github.com/libgit2/libgit2/blob/74c37c2a48756299306bb10e9e546c1627ca289f/include/git2/merge.h#L546), [rebase](https://github.com/libgit2/libgit2/blob/master/include/git2/rebase.h#L205),[reset](https://github.com/libgit2/libgit2/blob/master/include/git2/rebase.h#L205), [reset](https://github.com/libgit2/libgit2/blob/master/include/git2/rebase.h#L205), [revert](https://github.com/libgit2/libgit2/blob/62d38a1ddb8081f8b46e7ff6e21ebeb0014162d6/include/git2/revert.h#L33), [submodule](https://github.com/libgit2/libgit2/blob/62d38a1ddb8081f8b46e7ff6e21ebeb0014162d6/include/git2/submodule.h#L130).

Note that some of them have those options as a parameter, others have them included in another `git_*_options`.

tiennou commented:
Rebased that against master, and added support for the just-landed stash API.

Note that b4d58161908dc74d614999b43e3bd7a8707a95ee, fcbcf23a725e11eb21ff31fc980654383a0a4517 and cc4197685a6c74dd90a0d9f3543352d6110a00d6 have nothing to do with proposal, so cherry-picking them MIGHT be good.

joshaber commented:
:metal: 

Looks like some tests need to be updated still.

tiennou commented:
My bad, forgot to run the tests :-S. It should pass now.

joshaber commented:
Looks like there are still some legit build failures :disappointed:

tiennou commented:
Really, that thing was built, and tested here...

http://i.imgur.com/JESUH96.jpg

joshaber commented:
:clap:

tiennou commented:
Should be good now. Maybe. Who Knows.

joshaber commented on ObjectiveGit/GTRepository+Stashing.m:
```diff
@@ -68,29 +68,32 @@ - (BOOL)applyStashAtIndex:(NSUInteger)index flags:(GTRepositoryStashApplyFlag)fl
+		stash_options.checkout_options = *options.git_checkoutOptions;
+	}
+
+	if (pop == NO) {
```
This combining of the methods feels off to me since they're pretty different things.

tiennou replied:
Yeah, I wasn't really proud of that when I DRYed it... I'll revert that one.

joshaber commented on ObjectiveGit/GTRepository.h:
```diff
@@ -228,6 +203,9 @@ extern NSString * const GTRepositoryInitOptionsOriginURLString;
 /// returns nil (and fills the error parameter) if an error occurred, or a GTRepository object if successful.
++ (nullable instancetype)cloneFromURL:(NSURL *)originURL toWorkingDirectory:(NSURL *)workdirURL options:(nullable NSDictionary *)options error:(NSError **)error transferProgressBlock:(nullable void (^)(const git_transfer_progress *, BOOL *stop))transferProgressBlock;
+
+/// Backward-compatible method that uses `GTRepositoryCloneOptionsCheckout`
```
Do you think we should mark it deprecated too?

tiennou replied:
I actually started by keeping the old methods "checkout-ing" methods in (this one and the `-checkoutStuff:` below). Then it started to get messy with keeping the old "no notify" methods, so they were slashed, but this one stayed in. If you don't mind about bc, I can :fire: it.

joshaber replied:
I think I'd prefer we deprecate them :+1:

joshaber replied:
It looks like these still need to be deprecated.

joshaber commented on ObjectiveGit/GTRepository.m:
```diff
@@ -293,8 +305,6 @@ + (instancetype)cloneFromURL:(NSURL *)originURL toWorkingDirectory:(NSURL *)work
 
 	return [[self alloc] initWithGitRepository:repository];
-
-	return nil;
```
Ha! Nice catch.

tiennou replied:
Yeah, I'm not sure why Xcode didn't complain about dead-code-whatever ;-).

joshaber commented on ObjectiveGit/GTCheckoutOptions.m:
```diff
@@ -0,0 +1,97 @@
+static void GTCheckoutProgressCallback(const char *path, size_t completedSteps, size_t totalSteps, void *payload) {
+	if (payload == NULL) return;
+	void (^block)(NSString *, NSUInteger, NSUInteger) = (__bridge id)payload;
+	NSString *nsPath = (path != NULL ? [NSString stringWithUTF8String:path] : nil);
```
Style: could use `@(path)` instead of `+stringWithUTF8String:`.

joshaber commented on ObjectiveGit/GTCheckoutOptions.m:
```diff
@@ -0,0 +1,97 @@
+- (git_checkout_options *)git_checkoutOptions {
+	_git_checkoutOptions.checkout_strategy = self.strategy;
+
+	if (self.progressBlock) {
```
Style: use an explicit comparison to `nil`.

joshaber commented on ObjectiveGit/GTCheckoutOptions.m:
```diff
@@ -0,0 +1,97 @@
+		_git_checkoutOptions.progress_payload = (__bridge void *)self.progressBlock;
+	}
+
+	if (self.notifyBlock) {
```
Same as above about comparing to `nil`.

joshaber commented:
:metal: 

Some :memo: above.

tiennou commented on ObjectiveGit/GTRepository.h:
```diff
@@ -216,7 +191,7 @@ extern NSString * const GTRepositoryInitOptionsOriginURLString;
 ///                         `GTRepositoryCloneOptionsTransportFlags`,
 ///                         `GTRepositoryCloneOptionsBare`,
-///                         `GTRepositoryCloneOptionsCheckout`,
+///                         `GTRepositoryCloneCheckoutOptions`,
```
Bikeshedding welcome ;-).

tiennou commented on ObjectiveGit/GTCheckoutOptions.h:
```diff
@@ -0,0 +1,63 @@
+@property (copy) void (^progressBlock)(NSString *path, NSUInteger completedSteps, NSUInteger totalSteps);
+
+@property (assign) GTCheckoutNotifyFlags notifyFlags;
+@property (copy) int (^notifyBlock)(GTCheckoutNotifyFlags why, NSString *path, GTDiffFile *baseline, GTDiffFile *target, GTDiffFile *workdir);
```
Also, the copy-ness here makes me twitchy.

pietbrauer commented:
@joshaber @tiennou Whats the state of this one? Seems, comments were applied, tests run through and everything seems fine 🤔

tiennou commented:
Well, apart from my own questions on naming things and copying blocks, it still looks fine ;-).

joshaber commented:
Ah, I didn't realize it was ready for re-review :smile:

joshaber commented:
Just :one: :notebook_with_decorative_cover: about deprecating the old methods.

tiennou commented:
Rebased because Xcode decided to eat my submodules (thanks Xcode). Sorry about that. I've removed all stuff related to my tentative merging of the stash machinery since that was unrelated and added 8449a34 to handle pathspecs.

@joshaber I can't seem to parse your emojies, what did you meant ?

joshaber commented:
Ah, sorry @tiennou that was supposed to be "one note about deprecating old methods." 😄 I commented again above.

I don't do much with ObjectiveGit these days, so it'd be cool if @pietbrauer could weigh in.

tiennou commented:
Deprecated methods got the 🔥. I haven't been able to run tests here because of a linking failure with http-parser (I think, I can't even find how libgit2 gets linked in there...). I *think* that, because I have that brewed, the build process uses it instead of the internal one, which then shows up as a link failure here.

pietbrauer commented:
@tiennou Could you rebase onto master so the tests run?

tiennou commented:
Rebased, but 1) the iOS failed, 2) it doesn't look like it ran tests against a6dce75.

tiennou commented:
It seems the iOS tests are broken ?

pietbrauer commented:
@tiennou I guess they are a bit flaky. I don't know what this Quick message even does, sometimes it pops up, `push` build went through, though.

tiennou commented:
@pietbrauer Can this be reviewed/merged ? It seems @slavikus has made a few commits that add checkout customization support to stash, which was the point of this PR.

pietbrauer commented:
@tiennou Can you and @slavikus sort that out? I don't have a huge stake in this PR and just pinged people so we can merge/close some PRs.

tiennou commented:
I just am generally cautious about abusing my own merge rights, that's all. Let's say that if there's no movement here until next year, I'll merge.

slavikus commented:
@tiennou the new year is here, what should we do? :)

tiennou commented:
Added documentation and fixed a few dangling things. @slavikus can you review that ? I just want to make sure it's 👍  by at least another set of eyes before merging...

pietbrauer commented:
I felt adventurous today so I thought I would merge this. Looked good to me and it's not that the world would stop moving if we merge this. Thanks for the effort @tiennou!

slavikus commented:
Dangit, missed the initial notification back in January for some reason. I've been following this merge for a while, and I think it's looking good. Thanks guys!
